### ISSUE TITLE
Can't type in chat immediately after returning to lobby screen from gameplay unless you click with mouse

### ISSUE THREAD

**Issue by bdach** [ppy/osu#15716]
This actually looks to be related to framework-side things. The scenario here is:

* Multiplayer player is exited
* The multiplayer lobby chat textbox, being a `FocusedTextBox`, [eagerly takes keyboard focus](https://github.com/ppy/osu/blob/0641d5ddddd9919ee59f19a59b554387e5d586ad/osu.Game/Graphics/UserInterface/FocusedTextBox.cs#L99), therefore activating text input at SDL level [via `bindInput()`](https://github.com/ppy/osu-framework/blob/d7bb25be651d040054c533b63a35f78176e4081a/osu.Framework/Graphics/UserInterface/TextBox.cs#L1011-L1017)
* But the multi screen's own chat text box [drops keyboard input on disposal](https://github.com/ppy/osu-framework/blob/d7bb25be651d040054c533b63a35f78176e4081a/osu.Framework/Graphics/UserInterface/TextBox.cs#L350-L357) after that, and the lobby chat text box does not attempt to re-acquire focus again
* Therefore leading to a situation wherein the text box is activated, but the underlying singleton `ITextInputSource` is not, and therefore calling `GetPendingText()` returns an empty string.

I've confirmed this by applying the following framework-side patch and verifying that the assertion trips:

<details>
<summary>patch</summary>

```diff
diff --git a/osu.Framework/Input/SDL2DesktopWindowTextInput.cs b/osu.Framework/Input/SDL2DesktopWindowTextInput.cs
index fa3d8b74c1..c0edcf80a6 100644
--- a/osu.Framework/Input/SDL2DesktopWindowTextInput.cs
+++ b/osu.Framework/Input/SDL2DesktopWindowTextInput.cs
@@ -2,6 +2,7 @@
 // See the LICENCE file in the repository root for full licence text.
 
 using System;
+using System.Diagnostics;
 using osu.Framework.Platform;
 
 namespace osu.Framework.Input
@@ -20,6 +21,8 @@ public SDL2DesktopWindowTextInput(SDL2DesktopWindow window)
 
         public string GetPendingText()
         {
+            Trace.Assert(window.TextInputActive);
+
             try
             {
                 return pending;
diff --git a/osu.Framework/Platform/SDL2DesktopWindow.cs b/osu.Framework/Platform/SDL2DesktopWindow.cs
index a1abcaa869..49fc26564c 100644
--- a/osu.Framework/Platform/SDL2DesktopWindow.cs
+++ b/osu.Framework/Platform/SDL2DesktopWindow.cs
@@ -577,6 +577,8 @@ private void pollMouse()
 
         public void StopTextInput() => ScheduleCommand(SDL.SDL_StopTextInput);
 
+        public bool TextInputActive => SDL.SDL_IsTextInputActive() == SDL.SDL_bool.SDL_TRUE;
+
         #region SDL Event Handling
 
         /// <summary>
```

</details>

(cc: @Susko3)

### Discussed in https://github.com/ppy/osu/discussions/15715

<div type='discussions-op-text'>

<sup>Originally posted by **Theighlin** November 21, 2021</sup>
This used to work before

https://user-images.githubusercontent.com/53081750/142760414-61c45f49-3549-4465-a70a-af2a659eb541.mp4

</div>

**Comment by Susko3** [comment]
Just a quick question:
If a focused drawable is disposed, does it get unfocused first?

That `unbindInput()` probably shouldn't be there if that's the case.

`ITextInputSource` and `SDL2TextInputSource` are coded in a way they don't expect two textboxes to have focus, and also expect that for every `Activate()`, a `Deactivate()` will arrive (sooner or later).

`Dispose` currently unconditionally calls `unbindInput()`, causing a `Deactivate()` even if the text input isn't active.

**Comment by Susko3** [comment]
![image](https://user-images.githubusercontent.com/16479013/142763235-df7bd6e1-8758-4c16-bd05-eed5e4e4134a.png)
(Press `Ctrl+F` to see for yourself.)

Can we even rely on one focused textbox at a time?

Both textboxes here are receiving text input, but only the search one is handling key presses.

**Comment by bdach** [comment]
> If a focused drawable is disposed, does it get unfocused first?

I'm not sure that unfocus is guaranteed to happen before disposal across screens / when drawables are removed from the scene graph entirely. But the bigger issue is that the textbox is releasing text input on disposal even if it hasn't taken it, which is a ticking time bomb given async disposal as it can deactivate another textbox's active input.

> Both textboxes here are receiving text input, but only the search one is handling key presses.

The issue you mention is more likely related to `ManualInputManager` itself which will receive events with priority without handling and proxy them to its children, if I recall correctly. It's mostly a testing concern, I believe.

**Comment by Susko3** [comment]
> textbox is releasing text input on disposal even if it hasn't taken it

Should be easy enough to fix, just keep track of activation, and only unbind if active.

> which is a ticking time bomb given async disposal as it can deactivate another textbox's active input at any time

Now this is a bit harder.

To fix this, I propose the following:

Single-instance `ITextInputSource` (honestly makes a lot more sense than having one for every `TextBox`).
Then, when `ITextInputSource.Activate()` is called, increment a counter, same goes for `ITextInputSource.Deactivate()`, decrementing it.

Then only call the native functions for enable if counter goes from 0 to 1, and deactivate if it goes from 1 to 0. `Activate` incrementing past 1 would just call `EnsureActivated()`.

In the bug above, the counter would go 1 &rarr; 2 &rarr; 1, so input would never get deactivated.

Since we'd want this logic to apply to every `ITextInputSource`, I think we should move `ITextInputSource` to an abstract class and implement the logic there.

TL;DR keep an activation count counter in `ITextInputSource` and call native deactivate function only when it's 0.

If you're fine with this, I can bundle the two fixes in a PR. (Just to note, only the first fix is necessary to fix the bug here.)

**Comment by bdach** [comment]
I guess I could be ok with a counter as proposed. Not sure about the others.

On the other side text box shouldn't really be allowed to keep input active until its indeterminate disposal time, ideally it would deactivate input synchronously upon being removed from its parent. But I'm not sure there is a way to achieve that currently - it's a repeating pain point with async disposal (see: https://github.com/ppy/osu-framework/pull/4818, https://github.com/ppy/osu-framework/pull/3894, and probably more threads that I can't currently find). So maybe the counter is the best intermediate solution...
